pkg/urbit/vere/io/ames.c

```diff
@@ -446,38 +667,146 @@ _ames_recv_cb(uv_udp_t*        wax_u,
 {
   u3_ames* sam_u = wax_u->data;
 
-  //  data present, and protocol version in header matches 0
+  c3_o pas_o = c3y;
+
+  //  ensure a sane message size
   //
-  //    XX inflexible, scry version out of ames
+  if (4 >= nrd_i) {
+    pas_o = c3n;
+  }
+
+  //  ensure the protocol version matches ours
   //
-  if (  (0 < nrd_i)
-     && (0 == (0x7 & *((c3_w*)buf_u->base))) )
+  if ( c3y == pas_o
+    && (c3y == sam_u->fit_o)
+    && (sam_u->ver_y != (0x7 & *((c3_w*)buf_u->base))) )
   {
-    u3_noun wir = u3nc(c3__ames, u3_nul);
-    u3_noun cad;
+    pas_o = c3n;
+    //TODO  unless sender is our sponsee (transitively?)
+    //TODO  how does this interact with forwards?
 
-    {
-      u3_noun msg = u3i_bytes((c3_w)nrd_i, (c3_y*)buf_u->base);
-      u3_noun lan;
+    sam_u->vet_d++;
+    if ( 0 == (sam_u->vet_d % 100) ) {
+      u3l_log("ames: %" PRIu64 " dropped for version mismatch\n", sam_u->vet_d);
+    }
+  }
 
-      {
-        struct sockaddr_in* add_u = (struct sockaddr_in *)adr_u;
-        u3_lane             lan_u;
+  if (c3y == pas_o) {
+    c3_y*   byt_y = (c3_y*)buf_u->base;
+    c3_y*   bod_y = byt_y + 4;
+    u3_head hed_u;
 
-        lan_u.por_s = ntohs(add_u->sin_port);
-        lan_u.pip_w = ntohl(add_u->sin_addr.s_addr);
-        lan = u3_ames_encode_lane(lan_u);
-      }
+    //  unpack the packet header
+    //
+    {
+      c3_w hed_w = (byt_y[0] <<  0)
+                 | (byt_y[1] <<  8)
+                 | (byt_y[2] << 16)
+                 | (byt_y[3] << 24);
+
+      hed_u.ver_y = hed_w & 0x7;
+      hed_u.mug_l = (hed_w >> 3) & ((1 << 20) - 1);
+      hed_u.sac_y = (hed_w >> 23) & 0x3;
+      hed_u.rac_y = (hed_w >> 25) & 0x3;
+      hed_u.enc_o = (hed_w >> 26) & 0x1;
+    }
 
-      cad = u3nt(c3__hear, u3nc(c3n, lan), msg);
+    //  ensure the mug is valid
+    //
+    if ( hed_u.mug_l != _ca_mug_body(nrd_i - 4, bod_y) ) {
+      pas_o = c3n;
+      sam_u->mut_d++;
+      if ( 0 == (sam_u->mut_d % 100) ) {
+        u3l_log("ames: %" PRIu64 " dropped for invalid mug\n", sam_u->mut_d);
+      }
     }
+    //  if we can scry, we might want to forward statelessly
+    //
+    else if (c3y == sam_u->see_o) {
+      c3_y sen_y = 2 << hed_u.sac_y;
+      c3_y rec_y = 2 << hed_u.rac_y;
 
-    u3_auto_peer(
-      u3_auto_plan(&sam_u->car_u,
-                   u3_ovum_init(0, c3__a, wir, cad)),
-      0, 0, _ames_hear_bail);
+      u3_noun sen = u3i_bytes(sen_y, bod_y);
+      u3_noun rec = u3i_bytes(rec_y, bod_y + sen_y);
```

Comment:
I think these get leaked if we're not forwarding.  This won't affect stars, galaxies, or the half of planets which are direct atoms, but it will moons and comets.